nixos/modules/services/web-apps/librechat.nix

```diff
@@ -0,0 +1,232 @@
+      }
+      {
+        assertion =
+          (cfg.env ? CREDS_KEY || cfg.credentials ? CREDS_KEY)
+          && (cfg.env ? CREDS_IV || cfg.credentials ? CREDS_IV)
+          && (cfg.env ? JWT_SECRET || cfg.credentials ? JWT_SECRET)
+          && (cfg.env ? JWT_REFRESH_SECRET || cfg.credentials ? JWT_REFRESH_SECRET);
+        message = ''
+          CREDS_KEY, CREDS_IV, JWT_SECRET, and JWT_REFRESH_SECRET must be defined in `services.librechat.credentials` and point to locations of files on the host.
+          Alternatively it can be defined in `services.librechat.env` with literal values but they will be saved within the world-readable nix store.;
+          You can use https://www.librechat.ai/toolkit/creds_generator to generate these.
+        '';
+      }
+    ];
+
+    networking.firewall.allowedTCPPorts = lib.optional cfg.openFirewall cfg.port;
+
+    systemd.tmpfiles.settings."10-librechat"."${cfg.dataDir}".d = {
+      mode = "0755";
+      inherit (cfg) user group;
+    };
+
+    systemd.services.librechat = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "tmpfiles.target" ];
+      description = "Open-source app for all your AI conversations, fully customizable and compatible with any AI provider";
+      environment = cfg.env;
+      script = # sh
+        ''
+          ${exportAllCredentials cfg.credentials}
+          cd ${cfg.dataDir}
+          ${lib.getExe cfg.package}
+        '';
+      serviceConfig = {
+        Type = "simple";
+        User = cfg.user;
+        Group = cfg.group;
+        StateDirectory = builtins.baseNameOf cfg.dataDir;
+        WorkingDirectory = cfg.dataDir;
+        LoadCredential = getLoadCredentialList;
+        Restart = "on-failure";
+        RestartSec = 10;
+
+        # Hardening
+        CapabilityBoundingSet = "";
+        NoNewPrivileges = true;
+        PrivateUsers = true;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        PrivateMounts = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        RestrictAddressFamilies = [
+          "AF_INET"
+          "AF_INET6"
+          "AF_UNIX"
+        ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        UMask = "0077";
+      };
+    };
+
+    users.users.librechat = lib.mkIf (cfg.user == "librechat") {
+      name = "librechat";
+      isSystemUser = true;
+      group = "librechat";
+      description = "LibreChat server user";
+    };
+
+    users.groups.librechat = lib.mkIf (cfg.user == "librechat") { };
+
+    # TODO: can we get the port using nix? where is the port even defined, what if the user change it?
+    services.librechat.env.MONGO_URI = lib.mkIf cfg.enableLocalDB "mongodb://localhost:27017";
```

Comment:
> can we get the port using nix?

You should be able to if you use ferretdb instead, it seems the option is exposed.

Reply:
> I started working on it but it's not functional yet, will get back to this in a few hours.

This was caused by a typo, I gave it an invalid port number.

However ferretdb itself is not working (yet) for librechat: `findAndModify: support for field "fields" with value &{map[_id:1] [{1 _id}] false 0} is not implemented yet`

I think this should be good after a version bump, I saw some `findAndModify` tests upstream.

For now I'll revert the mongodb -> ferretdb change until it is resolved (https://github.com/NixOS/nixpkgs/issues/436486).

Later it could be useful as a starting point when we can use ferretdb: 9cece9d78214

Reply:
Yes, I've been following your work related to this recently, I wish you the best and I'll be sure to switch use FerretDB once LibreChat works with it :)

I don't know much about postgres extensions, so I can't help with those issues or reviews.